server/src/main/java/org/elasticsearch/index/mapper/JsonFieldMapper.java

```diff
@@ -0,0 +1,321 @@
+
+        public void setSplitQueriesOnWhitespace(boolean splitQueriesOnWhitespace) {
+            checkIfFrozen();
+            this.splitQueriesOnWhitespace = splitQueriesOnWhitespace;
+        }
+
+        @Override
+        public Query existsQuery(QueryShardContext context) {
+            if (omitNorms()) {
+                return new TermQuery(new Term(FieldNamesFieldMapper.NAME, name()));
+            } else {
+                return new NormsFieldExistsQuery(name());
+            }
+        }
+
+        @Override
+        public Query fuzzyQuery(Object value, Fuzziness fuzziness, int prefixLength, int maxExpansions,
+                                boolean transpositions) {
+            throw new UnsupportedOperationException("[fuzzy] queries are not currently supported on [" +
+                CONTENT_TYPE + "] fields.");
+        }
+
+        @Override
+        public Query regexpQuery(String value, int flags, int maxDeterminizedStates,
+                                 MultiTermQuery.RewriteMethod method, QueryShardContext context) {
+            throw new UnsupportedOperationException("[regexp] queries are not currently supported on [" +
+                CONTENT_TYPE + "] fields.");
+        }
+
+        @Override
+        public Object valueForDisplay(Object value) {
+            if (value == null) {
+                return null;
+            }
+            BytesRef binaryValue = (BytesRef) value;
+            return binaryValue.utf8ToString();
+        }
+    }
+
+    private final JsonFieldParser fieldParser;
+    private int ignoreAbove;
+
+    private JsonFieldMapper(String simpleName,
+                            MappedFieldType fieldType,
+                            MappedFieldType defaultFieldType,
+                            int ignoreAbove,
+                            Settings indexSettings) {
+        super(simpleName, fieldType, defaultFieldType, indexSettings, MultiFields.empty(), CopyTo.empty());
+        assert fieldType.indexOptions().compareTo(IndexOptions.DOCS_AND_FREQS) <= 0;
+
+        this.ignoreAbove = ignoreAbove;
+        this.fieldParser = new JsonFieldParser(fieldType, ignoreAbove);
+    }
+
+    @Override
+    protected String contentType() {
+        return CONTENT_TYPE;
+    }
+
+    @Override
+    protected void doMerge(Mapper mergeWith) {
+        super.doMerge(mergeWith);
+        this.ignoreAbove = ((JsonFieldMapper) mergeWith).ignoreAbove;
+    }
+
+    @Override
+    protected JsonFieldMapper clone() {
+        return (JsonFieldMapper) super.clone();
+    }
+
+    @Override
+    public JsonFieldType fieldType() {
+        return (JsonFieldType) super.fieldType();
+    }
+
+    @Override
+    protected void parseCreateField(ParseContext context, List<IndexableField> fields) throws IOException {
+        if (context.parser().currentToken() == XContentParser.Token.VALUE_NULL) {
+            return;
+        }
+
+        if (fieldType().indexOptions() != IndexOptions.NONE || fieldType().stored()) {
```

Comment:
I think the stored check here is obsolete since we aren't adding an option for a stored field as per https://github.com/elastic/elasticsearch/pull/33923/files#r219319713 below? We should probably also throw an exception if `store(boolean)` is called on the `Builder`?

Also I wonder if its worth even allowing the `indexOptions` to be set? Given you can't turn on term vectors or positions the only option would be to turn off indexing altogether which I don't think makes sense since we aren't allowing stored fields? Whats left would be the user wanting to ignore the field completely which is already covered by the `object` type's `enabled" false` setting?

Reply:
The lucene fields we're adding are still stored if `stored: true`, it's just that each value is added as a separate field (as opposed to adding a single stored field containing the whole JSON blob, as I was brainstorming). I think it'd be nice to see how this naive approach looks on the feature branch, so we can get a sense of whether it looks confusing in kibana, etc. and make a choice about whether we should switch to a single stored field.

Reply:
Your understanding of the current behavior for stored fields is correct -- in this PR, a separate stored field is created for each leaf value. I think that later we should switch to the approach where we create a single stored field with the whole JSON blob (and have made a note of this on the meta-issue), but I was curious to play around with this first more naive approach. I'm also okay disallowing stored fields for now until we address the issue it in a dedicated PR.

> it wouldn't really make sense for a search for e.g. "json_field.a": "foo" since we wouldn't have the frequencies for the inner field alone

I think we would have these frequencies, because we'd calculate the frequency of the term `a\0foo` (where `\0` is a separator character)? And you're right, we are supporting both of those cases.

Reply:
> I think we would have these frequencies, because we'd calculate the frequency of the term a\0foo (where \0 is a separator character)?

Good point. In which case I agree that index options do make sense.

Reply:
After catching up with @romseygeek and @colings86 offline, I decided to disable stored fields in this PR to cut down on confusion.